File: src/js/helpers/schema.ts
import { z } from "zod";
import { Stream } from "node:stream";
import {
  LATEST_PROTOCOL_VERSION,
  SUPPORTED_PROTOCOL_VERSIONS,
} from "@modelcontextprotocol/sdk/types.js";

// Define the IOType enum values
const IOTypeSchema = z.union([
  z.literal("pipe"),
  z.literal("inherit"),
  z.literal("ignore"),
  z.instanceof(Stream),
  z.number(),
]);

const McpServerConfigurationSchema = z
  .object({
    protocolVersion: z
      .enum(SUPPORTED_PROTOCOL_VERSIONS as [string, ...string[]])
      .default(LATEST_PROTOCOL_VERSION)
      .describe("Protocol version"),
    jsonrpc: z.literal("2.0").default("2.0").describe("JSON-RPC version"),
    type: z.enum(["stdio"]).default("stdio"),
    command: z.string().describe("Command to execute (stdio only)"),
    args: z
      .array(z.string())
      .optional()
      .describe("Command arguments (stdio only)"),
    cwd: z.string().optional().describe("Working directory (stdio only)"),
    stderr: z
      .union([IOTypeSchema, z.instanceof(Stream), z.number()])
      .optional()
      .describe("How to handle stderr of the child process (stdio only)"),
    env: z
      .record(z.string(), z.string())
      .optional()
      .describe("Environment variables"),
    version: z.string().optional().describe("Version of the MCP server"),
  })
  .describe("Detected MCP server with analyzed configuration");

const CustomRequestMetaSchema = z
  .object({
    server: McpServerConfigurationSchema,
  })
  .passthrough();

const TraceSpanSchema = z.object({
  seq: z
    .string()
    .describe("Sequential identifier or name of the span operation"),
  parentSeq: z
    .string()
    .nullable()
    .describe("Sequential identifier of the parent span, null for root spans"),
  startTime: z
    .number()
    .describe("Start timestamp in milliseconds since Unix epoch"),
  duration: z
    .number()
    .optional()
    .describe("Duration of the span in milliseconds"),
  status: z
    .enum(["running", "success", "error"])
    .describe("Current status of the span execution"),
  error: z.string().optional().describe("Error message if the span failed"),
  data: z
    .record(z.string(), z.any())
    .optional()
    .describe("Additional custom data attached to the span"),
  isValid: z
    .boolean()
    .describe("Whether the span data is valid according to schema"),
});

const TraceSchema = z.object({
  traceId: z.string().describe("Unique identifier for the entire trace"),
  startTime: z.date().describe("Start time of the trace as a Date object"),
  endTime: z
    .date()
    .optional()
    .describe("End time of the trace as a Date object"),
  data: z
    .record(z.string(), z.any())
    .optional()
    .describe("Additional custom data attached to the trace"),
  spans: z
    .array(TraceSpanSchema)
    .describe("Array of all spans that occurred during this trace"),
  isValid: z
    .boolean()
    .describe("Whether the trace data is valid according to schema"),
});

const CustomErrorResponseSchema = z.object({
  code: z.number(),
  message: z.string(),
  data: z.record(z.any()).optional(),
});

export {
  McpServerConfigurationSchema,
  CustomRequestMetaSchema,
  TraceSpanSchema,
  TraceSchema,
  CustomErrorResponseSchema,
};

type McpServerConfiguration = z.infer<typeof McpServerConfigurationSchema>;
type CustomRequestMeta = z.infer<typeof CustomRequestMetaSchema>;
type TraceSpan = z.infer<typeof TraceSpanSchema>;
type Trace = z.infer<typeof TraceSchema>;
type CustomErrorResponse = z.infer<typeof CustomErrorResponseSchema>;

export type {
  McpServerConfiguration,
  CustomRequestMeta,
  TraceSpan,
  Trace,
  CustomErrorResponse,
};

File: src/python/app/helpers/schema.py
"""Schema definitions for MCP Outlet Python implementation.

This module provides Pydantic schemas that mirror the Zod schemas
in the TypeScript implementation, ensuring exact same input/output structure.
"""

from typing import Dict, Any, Optional, List, Union, Literal
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime


class McpServerConfiguration(BaseModel):
    """MCP Server Configuration schema (stdio only)."""

    protocol_version: str = Field(
        default="2025-03-26", description="Protocol version", alias="protocolVersion"
    )
    jsonrpc: Literal["2.0"] = Field(default="2.0", description="JSON-RPC version")
    type: Literal["stdio"] = Field(default="stdio")
    command: str = Field(description="Command to execute (stdio only)")
    args: Optional[List[str]] = Field(
        default=None, description="Command arguments (stdio only)"
    )
    cwd: Optional[str] = Field(
        default=None, description="Working directory (stdio only)"
    )
    stderr: Optional[Union[str, int]] = Field(
        default=None,
        description="How to handle stderr of the child process (stdio only)",
    )
    env: Optional[Dict[str, str]] = Field(
        default=None, description="Environment variables"
    )
    version: Optional[str] = Field(
        default=None, description="Version of the MCP server"
    )

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    def model_dump(self, **kwargs) -> Dict[str, Any]:
        """Override model_dump to ensure protocolVersion is used instead of protocol_version."""
        data = super().model_dump(by_alias=True, **kwargs)
        return data


class McpCallerConfiguration(BaseModel):
    """MCP Server Configuration schema (stdio only)."""

    protocol_version: str = Field(
        default="2025-03-26", description="Protocol version", alias="protocolVersion"
    )
    jsonrpc: Literal["2.0"] = Field(default="2.0", description="JSON-RPC version")
    type: Literal["stdio"] = Field(default="stdio")
    args: Optional[List[str]] = Field(
        default=None, description="Command arguments (stdio only)"
    )
    module_path: str = Field(description="Module path")
    package_name: str = Field(description="Module name")
    function_name: str = Field(description="Function name")

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    def model_dump(self, **kwargs) -> Dict[str, Any]:
        """Override model_dump to ensure protocolVersion is used instead of protocol_version."""
        data = super().model_dump(by_alias=True, **kwargs)
        return data


class CustomRequestMeta(BaseModel):
    """Request metadata schema."""

    server: McpServerConfiguration

    model_config = ConfigDict(extra="allow", populate_by_name=True)


class TraceSpan(BaseModel):
    """Trace span schema."""

    seq: str = Field(description="Sequential identifier or name of the span operation")
    parent_seq: Optional[str] = Field(
        default=None,
        description="Sequential identifier of the parent span, null for root spans",
        alias="parentSeq",
    )
    start_time: float = Field(
        description="Start timestamp in milliseconds since Unix epoch",
        alias="startTime",
    )
    duration: Optional[float] = Field(
        default=None, description="Duration of the span in milliseconds"
    )
    status: Literal["running", "success", "error"] = Field(
        description="Current status of the span execution"
    )
    error: Optional[str] = Field(
        default=None, description="Error message if the span failed"
    )
    data: Optional[Dict[str, Any]] = Field(
        default=None, description="Additional custom data attached to the span"
    )
    is_valid: bool = Field(
        description="Whether the span data is valid according to schema",
        alias="isValid",
    )

    model_config = ConfigDict(populate_by_name=True)


class Trace(BaseModel):
    """Trace schema."""

    trace_id: str = Field(
        description="Unique identifier for the entire trace", alias="traceId"
    )
    start_time: datetime = Field(
        description="Start time of the trace as a DateTime object", alias="startTime"
    )
    end_time: Optional[datetime] = Field(
        default=None,
        description="End time of the trace as a DateTime object",
        alias="endTime",
    )
    data: Optional[Dict[str, Any]] = Field(
        default=None, description="Additional custom data attached to the trace"
    )
    spans: List[TraceSpan] = Field(
        description="Array of all spans that occurred during this trace"
    )
    is_valid: bool = Field(
        description="Whether the trace data is valid according to schema",
        alias="isValid",
    )

    model_config = ConfigDict(populate_by_name=True)


class CustomErrorResponse(BaseModel):
    """Custom error response schema."""

    code: int
    message: str
    data: Optional[Dict[str, Any]] = None


# Supported protocol versions (mirrors TypeScript constants)
SUPPORTED_PROTOCOL_VERSIONS = ["2025-03-26", "2024-11-05", "2024-10-07"]
LATEST_PROTOCOL_VERSION = "2025-03-26"
